unbalanced algorithm for dynamicDivideReplicas

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

-->


#### What happened:
kind/bug

#### What you expected to happen:

By adding a testcase in `pkg/scheduler/framework/plugins/defaultassigner/algorithm_test.go` to test the algorithm on dividing replicas.

```go
// pkg/scheduler/framework/plugins/defaultassigner/algorithm_test.go
func TestDynamicDivideReplicas(t *testing.T) {
	tests := []struct {
		name                 string
		desiredReplicas      int32
		maxAvailableReplicas []int32
		want                 []int32
	}{
		{
			desiredReplicas:      2,
			maxAvailableReplicas: []int32{3, 6, 9},
			want:                 []int32{0, 1, 1}, // but  we got [1 0 1]
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := dynamicDivideReplicas(tt.desiredReplicas, tt.maxAvailableReplicas); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("dynamicDivideReplicas() = %v, want %v", got, tt.want)
			}
		})
	}
}
```

```
=== RUN   TestDynamicDivideReplicas
=== RUN   TestDynamicDivideReplicas/#00
    algorithm_test.go:448: dynamicDivideReplicas() = [1 0 1], want [0 1 1]
--- FAIL: TestDynamicDivideReplicas (0.00s)
    --- FAIL: TestDynamicDivideReplicas/#00 (0.00s)
```

When `desiredReplicas` is much smaller, the algorithm gets unbalanced.


#### How to reproduce it (as minimally and precisely as possible):

#### Anything else we need to know?:

#### Environment:
- Clusternet version:
    - Clusternet-agent version (user `clusternet-agent --version=json`):
    - Clusternet-hub version (user `clusternet-hub --version=json`):
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Others:

--- a/pkg/scheduler/framework/plugins/defaultassigner/algorithm.go
+++ b/pkg/scheduler/framework/plugins/defaultassigner/algorithm.go
@@ -1,6 +1,7 @@
 package defaultassigner
 
 import (
+	"math"
 	"sort"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -94,7 +95,7 @@ func dynamicDivideReplicas(desiredReplicas int32, maxAvailableReplicas []int32) 
 
 	type cluster struct {
 		index   int
-		decimal float32
+		decimal float64
 	}
 	clusters := make([]cluster, 0, len(maxAvailableReplicas))
 
@@ -104,7 +105,7 @@ func dynamicDivideReplicas(desiredReplicas int32, maxAvailableReplicas []int32) 
 		remain -= replica
 		clusters = append(clusters, cluster{
 			index:   i,
-			decimal: float32(weight*desiredReplicas)/float32(weightSum) - float32(replica),
+			decimal: math.Abs(float64(weight*desiredReplicas)/float64(weightSum) - float64(replica)),
 		})
 	}
 

--- a/pkg/scheduler/framework/plugins/defaultassigner/algorithm.go
+++ b/pkg/scheduler/framework/plugins/defaultassigner/algorithm.go
@@ -91,19 +91,35 @@ func dynamicDivideReplicas(desiredReplicas int32, maxAvailableReplicas []int32) 
 	}
 
 	remain := desiredReplicas
+
+	type cluster struct {
+		index   int
+		decimal float32
+	}
+	clusters := make([]cluster, 0, len(maxAvailableReplicas))
+
 	for i, weight := range maxAvailableReplicas {
 		replica := weight * desiredReplicas / weightSum
 		res[i] = replica
 		remain -= replica
+		clusters = append(clusters, cluster{
+			index:   i,
+			decimal: float32(weight*desiredReplicas)/float32(weightSum) - float32(replica),
+		})
 	}
+
+	// sort the clusters by descending order of decimal part of replica
+	sort.Slice(clusters, func(i, j int) bool {
+		return clusters[i].decimal > clusters[j].decimal
+	})
 
 	if remain > 0 {
 		for i := 0; i < int(remain) && i < len(res); i++ {
-			res[i]++
+			res[clusters[i].index]++
 		}
 	} else if remain < 0 {
 		for i := 0; i < int(-remain) && i < len(res); i++ {
-			res[i]--
+			res[clusters[i].index]--
 		}
 	}
 

--- a/pkg/scheduler/framework/plugins/defaultassigner/algorithm_test.go
+++ b/pkg/scheduler/framework/plugins/defaultassigner/algorithm_test.go
@@ -427,3 +427,45 @@ func TestDivideReplicas(t *testing.T) {
 		})
 	}
 }
+
+func Test_dynamicDivideReplicas(t *testing.T) {
+	tests := []struct {
+		name                 string
+		desiredReplicas      int32
+		maxAvailableReplicas []int32
+		want                 []int32
+	}{
+		{
+			desiredReplicas:      6,
+			maxAvailableReplicas: []int32{3, 6, 9},
+			want:                 []int32{1, 2, 3},
+		},
+		{
+			desiredReplicas:      2,
+			maxAvailableReplicas: []int32{3, 6, 9},
+			want:                 []int32{0, 1, 1},
+		},
+		{
+			desiredReplicas:      3,
+			maxAvailableReplicas: []int32{3, 6, 9},
+			want:                 []int32{1, 1, 1},
+		},
+		{
+			desiredReplicas:      3,
+			maxAvailableReplicas: []int32{0, 0, 0},
+			want:                 []int32{0, 0, 0},
+		},
+		{
+			desiredReplicas:      6,
+			maxAvailableReplicas: []int32{1, 2, 2},
+			want:                 []int32{1, 2, 2},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := dynamicDivideReplicas(tt.desiredReplicas, tt.maxAvailableReplicas); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("dynamicDivideReplicas() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}

--- a/pkg/scheduler/framework/plugins/defaultassigner/algorithm_test.go
+++ b/pkg/scheduler/framework/plugins/defaultassigner/algorithm_test.go
@@ -460,6 +460,16 @@ func Test_dynamicDivideReplicas(t *testing.T) {
 			maxAvailableReplicas: []int32{1, 2, 2},
 			want:                 []int32{1, 2, 2},
 		},
+		{
+			desiredReplicas:      -6,
+			maxAvailableReplicas: []int32{3, 6, 9},
+			want:                 []int32{-1, -2, -3},
+		},
+		{
+			desiredReplicas:      -2,
+			maxAvailableReplicas: []int32{3, 6, 9},
+			want:                 []int32{0, -1, -1},
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

